graph_weather/models/atmorep/utils/uncertainity.py

```diff
@@ -0,0 +1,467 @@
+            ensemble_preds: Ensemble predictions with shape [E, B, T, H, W]
+            num_bins: Number of bins for discretization
+
+        Returns:
+            Entropy map with shape [B, T, H, W]
+        """
+        # For continuous variables, we discretize into bins
+        E, B, T, H, W = ensemble_preds.shape
+
+        # Reshape for binning
+        flat_preds = ensemble_preds.reshape(E, -1)
+
+        # Find global min and max for consistent binning
+        global_min = flat_preds.min()
+        global_max = flat_preds.max()
+        bin_width = (global_max - global_min) / num_bins
+
+        # Initialize entropy map
+        entropy = torch.zeros(B, T, H, W)
+
+        # Calculate entropy for each pixel location
+        for b in range(B):
+            for t in range(T):
+                for h in range(H):
+                    for w in range(W):
+                        # Get ensemble predictions for this location
+                        pixel_preds = ensemble_preds[:, b, t, h, w]
+
+                        # Create histogram
+                        hist = torch.histc(
+                            pixel_preds, bins=num_bins, min=global_min, max=global_max
+                        )
+                        hist = hist / E  # Normalize to get probability
+
+                        # Calculate entropy (avoiding log(0))
+                        valid_probs = hist[hist > 0]
+                        pixel_entropy = -torch.sum(valid_probs * torch.log2(valid_probs))
+                        entropy[b, t, h, w] = pixel_entropy
+
+        return entropy
+
+    def optimized_entropy(self, ensemble_preds, num_bins=10):
+        """
+        Vectorized implementation of entropy calculation for better performance
+
+        Args:
+            ensemble_preds: Ensemble predictions with shape [E, B, T, H, W]
+            num_bins: Number of bins for discretization
+
+        Returns:
+            Entropy map with shape [B, T, H, W]
+        """
+        E, B, T, H, W = ensemble_preds.shape
+
+        # Find global min and max for consistent binning
+        global_min = ensemble_preds.min()
+        global_max = ensemble_preds.max()
+
+        # Reshape to [E, B*T*H*W]
+        reshaped_preds = ensemble_preds.reshape(E, -1)
+
+        # Compute entropy for each spatial-temporal location
+        entropy = torch.zeros(B * T * H * W)
+
+        for i in range(B * T * H * W):
+            # Get predictions for this location
+            pixel_preds = reshaped_preds[:, i]
+
+            # Create histogram
+            hist = torch.histc(pixel_preds, bins=num_bins, min=global_min, max=global_max)
+            hist = hist / E  # Normalize
+
+            # Calculate entropy
+            valid_probs = hist[hist > 0]
+            pixel_entropy = -torch.sum(valid_probs * torch.log2(valid_probs))
+            entropy[i] = pixel_entropy
+
+        # Reshape back to [B, T, H, W]
+        return entropy.reshape(B, T, H, W)
+
+
+class CalibrationMetrics:
```

Comment:
These aren't bad, but I would maybe generalize these rank histograms and CRPS to a more general `losses/` or the `losses.py` file